### PR TITLE
Implement gRPC data handlers and SLAM thread.

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: cartographer_grpc/handlers/add_imu_data_handler.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+class AddImuDataHandler
+    : public framework::RpcHandler<framework::Stream<proto::AddImuDataRequest>,
+                                   google::protobuf::Empty> {
+ public:
+  void OnRequest(const proto::AddImuDataRequest &request) override {
+    // Note: The 'BlockingQueue' returned by 'sensor_data_queue()' is already
+    // thread-safe. Therefore it suffices to get an unsynchronized reference to
+    // the 'MapBuilderContext'.
+    GetUnsynchronizedContext<MapBuilderServer::MapBuilderContext>()
```
Comment (gaschler) [inline cartographer_grpc/handlers/add_imu_data_handler.h]: This seems to be a common pattern, perhaps move parts of it to a
MapBuilderContext::EnqueueSensorData(trajectory_id, ...)

File: cartographer_grpc/map_builder_server.cc [hunk 2]
```diff
@@ -13,15 +13,39 @@
...
+MapBuilderServer::MapBuilderContext::MapBuilderContext(
+    cartographer::mapping::MapBuilder* map_builder,
+    cartographer::common::BlockingQueue<SensorData>* sensor_data_queue)
+    : map_builder_(map_builder), sensor_data_queue_(sensor_data_queue) {}
+
+cartographer::mapping::MapBuilder&
+MapBuilderServer::MapBuilderContext::map_builder() {
+  return *map_builder_;
+}
+cartographer::common::BlockingQueue<MapBuilderServer::SensorData>&
```
Comment (gaschler) [inline cartographer_grpc/map_builder_server.cc]: empty line

Comment (cschuet) [reply cartographer_grpc/map_builder_server.cc]: Done.

File: cartographer_grpc/map_builder_server.h [hunk 3]
```diff
@@ -28,18 +28,27 @@ namespace cartographer_grpc {
...
 
    private:
-    cartographer::mapping::MapBuilder* map_builder_;
+    cartographer::mapping::MapBuilder *map_builder_;
+    cartographer::common::BlockingQueue<SensorData> *sensor_data_queue_;
   };
 
   MapBuilderServer(
-      const proto::MapBuilderServerOptions& map_builder_server_options);
+      const proto::MapBuilderServerOptions &map_builder_server_options);
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/map_builder_server.h]: Yeah, sorry this is when I work from my MacBook. Please ignore. I will try and fix this. But don't worry - Wally will clang-format before the PR is submitted :)

File: cartographer_grpc/map_builder_server.h [hunk 4]
```diff
@@ -28,18 +28,27 @@ namespace cartographer_grpc {
...
    public:
-    MapBuilderContext(cartographer::mapping::MapBuilder* map_builder)
-        : map_builder_(map_builder) {}
-    cartographer::mapping::MapBuilder& map_builder() { return *map_builder_; }
+    MapBuilderContext(
+        cartographer::mapping::MapBuilder *map_builder,
+        cartographer::common::BlockingQueue<SensorData> *sensor_data_queue);
+    cartographer::mapping::MapBuilder &map_builder();
+    cartographer::common::BlockingQueue<SensorData> &sensor_data_queue();
+    void AddSensorData(const SensorData &sensor_data);
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/map_builder_server.h]: Hmm, I don't think so: this must be callable from `MapBuilderServer::ProcessSensorDataQueue()`.

File: cartographer_grpc/framework/rpc_handler.h [hunk 5]
```diff
@@ -53,6 +53,10 @@ class RpcHandler : public RpcHandlerInterface {
   ExecutionContext::Synchronized<T> GetContext() {
     return {execution_context_->lock(), execution_context_};
   }
+  template <typename T>
+  T* GetUnsynchronizedContext() {
+    return static_cast<T*>(execution_context_);
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer_grpc/framework/rpc_handler.h]: Ok, this is a good compromise. I agree that making this type-safe requires many changes.
I think the underlying issue here is that RpcHandler implementations have a good idea what their request and response types are, but not about the state they need (in other words, ExecutionContext). I'd like to reconsider this in a future PR.

File: cartographer_grpc/framework/execution_context.h [hunk 6]
```diff
@@ -30,6 +30,7 @@ namespace framework {
 // 'RpcHandler::GetContext<MyContext>()'.
 class ExecutionContext {
  public:
+  virtual ~ExecutionContext() = default;
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/execution_context.h]: Done.

File: cartographer_grpc/framework/execution_context.h [hunk 7]
```diff
@@ -30,6 +30,7 @@ namespace framework {
...
   // This non-movable, non-copyable class is used to broker access from various
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/execution_context.h]: Done.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**